### PR TITLE
OD-213 [Feature] Allow sorting with query parameters

### PR DESCRIPTION
**Pull request by YaroslavOvdii**
@sofiiakvasnevska 

## Issue
https://weboo.atlassian.net/browse/OD-213

## Description
We have added parsing query parameters to get sort order and sort field values if users want to save their sort order in the LFD component.

## Screenshots/screencasts
https://share.getcloudapp.com/KoulEBXr

## Backward compatibility
This change is fully backward compatible.

## Reviewers
@ivandevupp

### CODE REVIEW

File: js/utils.js [hunk 1]
```diff
@@ -1059,6 +1059,12 @@ Fliplet.Registry.set('dynamicListUtils', (function() {
       return options.records;
     }
 
+    // Saving sort parametrs so when users comes back to the page through a BACK navigation, the sort order is restored
```
Comment (squallstar) [inline js/utils.js:1062]: typo in "parameters"

File: js/query-parser.js [hunk 2]
```diff
@@ -38,6 +38,21 @@ Fliplet.Registry.set('dynamicListQueryParser', function() {
   });
   this.queryPreFilter = _(this.pvPreFilterQuery).size() > 0;
 
+  // We can sort only by one column that is why this syntax dosn't supports
```
Comment (squallstar) [inline js/query-parser.js:41]: typo, change to: "doesn't support"

File: js/query-parser.js [hunk 3]
```diff
@@ -38,6 +38,21 @@ Fliplet.Registry.set('dynamicListQueryParser', function() {
...
+  // ?dynamicListFilterColumn=Name,Age&dynamicListFilterOrder=asc
+  // Correct example is
+  // ?dynamicListFilterColumn=Name&dynamicListFilterOrder=asc
+  this.pvPreSortQuery = _.pickBy({
```
Comment (squallstar) [inline js/query-parser.js:45]: can you explain what these `pvPreSortQuery` and `queryPreSort` do?

**Review by squallstar — CHANGES_REQUESTED** [review]

File: js/query-parser.js (hunk 3 above)
Comment (YaroslavOvdii) [reply js/query-parser.js:45]: @squallstar it was made like a filter query above and actually, all that these two variables do is:
1. `pvSortQuery` - is gathering query parameters from the `Fliplet. Navigate.query` method.
2. `queryPreSort` - determinates if the query parameters that we are looking at in the url are present.

File: js/query-parser.js [hunk 4]
```diff
@@ -38,6 +38,21 @@ Fliplet.Registry.set('dynamicListQueryParser', function() {
   });
   this.queryPreFilter = _(this.pvPreFilterQuery).size() > 0;
 
+  // We can sort only by one column that is why this syntax doesn't support
+  // ?dynamicListFilterColumn=Name,Age&dynamicListFilterOrder=asc
```
Comment (tonytlwu) [inline js/query-parser.js:42]: There was a mistake in the spec. The query parameters should refer to `sort` not `filter`.

```suggestion
  // ?dynamicListSortColumn=Name,Age&dynamicListSortOrder=asc
```

File: js/query-parser.js (hunk 3 above)
Comment (tonytlwu) [reply js/query-parser.js:45]:
```suggestion
  this.pvSortQuery = _.pickBy({
```

I would ignore the "pre" part. Put this further down the code close to where the filter and search queries are.

"Prefilters" and "presorts" would have referred to filtering and sorting the data before rendering the LFD. That's not what the current

**Review by tonytlwu — CHANGES_REQUESTED** [review]

**Comment by YaroslavOvdii** [comment]
@tonytlwu added changes

**Review by tonytlwu — APPROVED** [review]
@sofiiakvasnevska This is ready for testing

Risk: High
Affected: All LFD instances
Verify: LFD lists are rendered, ordered and working correctly with query parameters

Squash and merge: Use PR title

**Pull request merged**